### ISSUE TITLE
Do not set link-local address of tun device

### ISSUE THREAD

**Issue by nagy** [yggdrasil-network/yggdrasil-go#1150]
If my understanding of Yggdrasil is correct, than the link-local address of the tun device does not serve any purpose.
We could disable this address and therefore the tunnel interface does not "claim" to be able to send on that address.

So going from this:

```
18: ygg0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 65535 qdisc fq_codel state UNKNOWN group default qlen 500
    link/none
    inet6 203:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx/7 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::xxxx:xxxx:xxxx:xxxx/64 scope link stable-privacy proto kernel_ll
       valid_lft forever preferred_lft forever
```

to this:

```
17: ygg0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 65535 qdisc fq_codel state UNKNOWN group default qlen 500
    link/none
    inet6 203:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx/7 scope global
       valid_lft forever preferred_lft forever
```

This makes the interface list a little bit cleaner in my opinion.
You can achieve the same effect imperatively with:

```
# ip -6 address flush dev ygg0 scope link
```

I have had success with this following patch. What do you think? If you want I can make a PR out of this.

<details>

<summary>Patch</summary>

```diff
diff --git a/src/tun/tun_linux.go b/src/tun/tun_linux.go
index 98d63db..1266fca 100644
--- a/src/tun/tun_linux.go
+++ b/src/tun/tun_linux.go
@@ -60,6 +60,17 @@ func (tun *TunAdapter) setupAddress(addr string) error {
 	if err := netlink.LinkSetUp(nlintf); err != nil {
 		return fmt.Errorf("failed to bring link up: %w", err)
 	}
+	addrs, err := netlink.AddrList(nlintf, netlink.FAMILY_V6)
+	if err != nil {
+		return fmt.Errorf("failed to list addresses: %w", err)
+	}
+	for _, addr := range addrs {
+		if addr.Scope == int(netlink.SCOPE_LINK) {
+			if err := netlink.AddrDel(nlintf, &addr); err != nil {
+				return fmt.Errorf("failed to delete link-local address: %w", err)
+			}
+		}
+	}
 	// Friendly output
 	tun.log.Infof("Interface name: %s", tun.Name())
 	tun.log.Infof("Interface IPv6: %s", addr)
```

</details>

**Comment by neilalexander** [comment]
The link-local address wouldn't be used for any routable traffic so I'm not really sure if this has any benefit.

**Comment by tribals** [comment]
> The link-local address wouldn't be used for any routable traffic so I'm not really sure if this has any benefit.

If it will not be used then it simply should not exist.

It pollutes "local" routing table on Linux, so potentially it can lead to troubles on local network as LL addresses are heavily used for things like NDP, MLD and the like.

```console
$ ip -6 route show table local
local fe80::4e:bda4:8512:4e86 dev wlp3s0 proto kernel metric 0 pref medium
local fe80::2018:83e8:4eca:b214 dev ygg0 proto kernel metric 0 pref medium
```

Although, better to not set it up in the first place rather than delete afterwards, but I'm not sure if it is possible as LLs are assigned by kernel.

**Comment by tribals** [comment]
It's me again. Indeed, it is possible.

This is controlled by `net.ipv6.conf.*.addr_gen_mode`, as described in [kernel's "IP sysctl"](https://docs.kernel.org/networking/ip-sysctl.html#proc-sys-net-ipv6-variables). And can be changed with `ip-link(8)`:

```console
# ip link set dev ygg0 addrgenmode none
```

One could try to setting this trough `/etc/sysctl.d/xx-yggdrasil.conf`, like this:

```
net.ipv6.conf.ygg0.addr_gen_mode = 1
```

Unfortunately, this will not work for Yggdrasil as there will be no `ygg0` interface until Yggdrasil router is started, and `addrgenmode` should be set before link is brought up, so there will be no timeframe to do so when Yggdrasil router already started.

The solution is really simple, as `yggdrasil-go` already uses https://github.com/vishvananda/netlink for setting up it's interface:

```diff
diff --git a/src/tun/tun_linux.go b/src/tun/tun_linux.go
index 29f437a..6f23cca 100644
--- a/src/tun/tun_linux.go
+++ b/src/tun/tun_linux.go
@@ -8,6 +8,7 @@ import (
        "fmt"

        "github.com/vishvananda/netlink"
+       "github.com/vishvananda/netlink/nl"
        wgtun "golang.zx2c4.com/wireguard/tun"
 )

@@ -56,6 +57,9 @@ func (tun *TunAdapter) setupAddress(addr string) error {
        if err := netlink.LinkSetMTU(nlintf, int(tun.mtu)); err != nil {
                return fmt.Errorf("failed to set link MTU: %w", err)
        }
+       if err := netlink.LinkSetIP6AddrGenMode(nlintf, nl.IN6_ADDR_GEN_MODE_NONE); err != nil {
+               return fmt.Errorf("failed to set IPv6 addrgenmode: %w", err)
+       }
        if err := netlink.LinkSetUp(nlintf); err != nil {
                return fmt.Errorf("failed to bring link up: %w", err)
        }
```

Basically, the same as what [in-kernel WireGuard does](https://elixir.bootlin.com/linux/v7.1.4/source/drivers/net/wireguard/device.c#L48):

```c
	if (dev_v6)
		dev_v6->cnf.addr_gen_mode = IN6_ADDR_GEN_MODE_NONE;
```

Would you accept a PR for that?
